pkg/store/cache/caching_bucket.go

```diff
@@ -26,92 +29,178 @@ import (
-	}
 
 	cb := &CachingBucket{
 		Bucket: b,
-		config: chunks,
-		cache:  c,
 		logger: logger,
 
-		requestedChunkBytes: promauto.With(reg).NewCounter(prometheus.CounterOpts{
-			Name: "thanos_store_bucket_cache_requested_chunk_bytes_total",
-			Help: "Total number of requested bytes for chunk data.",
-		}),
-		fetchedChunkBytes: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
-			Name: "thanos_store_bucket_cache_fetched_chunk_bytes_total",
-			Help: "Total number of fetched chunk bytes. Data from bucket is then stored to cache.",
-		}, []string{"origin"}),
-		refetchedChunkBytes: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
-			Name: "thanos_store_bucket_cache_refetched_chunk_bytes_total",
-			Help: "Total number of chunk bytes re-fetched from storage, despite being in cache already.",
-		}, []string{"origin"}),
-		objectSizeRequests: promauto.With(reg).NewCounter(prometheus.CounterOpts{
-			Name: "thanos_store_bucket_cache_objectsize_requests_total",
-			Help: "Number of object size requests for objects.",
-		}),
-		objectSizeHits: promauto.With(reg).NewCounter(prometheus.CounterOpts{
-			Name: "thanos_store_bucket_cache_objectsize_hits_total",
-			Help: "Number of object size hits for objects.",
-		}),
-	}
-
-	cb.fetchedChunkBytes.WithLabelValues(originBucket)
-	cb.fetchedChunkBytes.WithLabelValues(originCache)
-	cb.refetchedChunkBytes.WithLabelValues(originCache)
+		operationConfigs: map[string][]*operationConfig{},
+
+		requestedGetRangeBytes: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_store_bucket_cache_getrange_requested_bytes_total",
+			Help: "Total number of bytes requested via GetRange.",
+		}, []string{config}),
+		fetchedGetRangeBytes: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_store_bucket_cache_getrange_fetched_bytes_total",
+			Help: "Total number of bytes fetched because of GetRange operation. Data from bucket is then stored to cache.",
+		}, []string{"origin", config}),
+		refetchedGetRangeBytes: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_store_bucket_cache_getrange_refetched_bytes_total",
+			Help: "Total number of bytes re-fetched from storage because of GetRange operation, despite being in cache already.",
+		}, []string{"origin", config}),
+
+		operationRequests: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_store_bucket_cache_operation_requests_total",
+			Help: "Number of requested operations matching given config.",
+		}, []string{"operation", "config"}),
+		operationHits: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_store_bucket_cache_operation_hits_total",
+			Help: "Number of operations served from cache for given config.",
+		}, []string{"operation", "config"}),
+	}
 
 	return cb, nil
 }
 
+func (cb *CachingBucket) addOperationConfig(operationName, configName string, matcher func(string) bool, cache cache.Cache, opConfig interface{}) {
```

Comment:
It would be nice to actually have it via constructor. We can create some `cacheConfigs` type if need with this method (:

Reply:
so we can avoid panics

Reply:
I've refactored config-setup into separate object (now called `CachingBucketConfig`), but I kept panics :) They are protection for coding errors, not config errors.